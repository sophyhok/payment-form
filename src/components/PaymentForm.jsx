import React, { useState } from 'react';
import parse from 'html-react-parser';

import Button from './FormUI/Button';
import Input from './FormUI/Input';
import DropDown from './FormUI/Dropdown';
import Container from './FormUI/Container';
import paymentFormValidation from './FormValidation/paymentFormValidation';

import amexImg from '../assets/img/amex.png';
import visaImg from '../assets/img/visa.png';

const PaymentForm = () => {
  const monthsList = ['01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12'];
  const yearsList = ['2021', '2022', '2023', '2024', '2025', '2026', '2027', '2028', '2029', '2030', '2031'];
  const bStatus = { name: 'none', card: 'none', cvv: 'none', exp: 'none' };
  const initialValue = { name: '', card: '', cvv: '', expMonth: monthsList[0], expYear: yearsList[0] };

  const [validation, setValidation] = useState(true);
  const [errors, setErrors] = useState({});
  const [boxStatus, setBoxStatus] = useState(bStatus);
  const [formValue, setFormValue] = useState(initialValue);

  const handleSubmit = event => {
    event.preventDefault();

    // check validation, return as object
    const validateResults = paymentFormValidation(formValue);

    // if not empty, there is an error in validation
    if (validateResults.length > 0) {
      setValidation(false);
      setErrors(validateResults);

      // setBoxStatus for input props
      validateResults.forEach(e => {
        let [key] = Object.keys(e);
        if (key === 'name') {
          setBoxStatus(prev => ({ ...prev, name: 'danger' }))
        }
        else if (key === 'cardNumber') {
          setBoxStatus(prev => ({ ...prev, card: 'danger' }))
        }
        else if (key === 'cvv') {
          setBoxStatus(prev => ({ ...prev, cvv: 'danger' }))
        }
        else if (key === 'expiration') {
          setBoxStatus(prev => ({ ...prev, exp: 'danger' }))
        }
      });
    }
    // no errors in validation
    else {
      setValidation(true);
      alert("Success! You have submitted your card information");
    }
  }

  const formatErrors = () => {
    let message = '';
    errors.forEach(e => {
      let [key] = Object.keys(e);
      message += `<div>${e[key]}</div>`;
    })
    return parse(message); // convert string to html
  }

  const handleChange = (event, key) => {
    setBoxStatus(bStatus)
    event.preventDefault();
    let value = event.target.value;

    // simple card format
    if (key === 'card') {
      // add space for visa
      if (value[0] === '4') {
        value = value.replace(/[^\dA-Z]/g, '').replace(/(.{4})/g, '$1 ');
      }
      // add space for amex
      else if (value.slice(0, 2) === '34' || value.slice(0, 2) === '37') {
        if (value.length === 4) {
          value += " ";
        }
        else if (value.length === 11) {
          value += " ";
        }
      }
    }

    setFormValue(prev => ({ ...prev, [event.target.name]: value }))
  }

  return (
    <form name="payment-form">
      <Container gap="none" justify="center" align="center">
        <h3>Enter your Credit Card information</h3>
        {!validation && (
          <Container gap="none" fontColor="danger">{formatErrors()}</Container>
        )}
        <div><img alt="visa" src={visaImg} /><img alt="amex" src={amexImg} /></div>
        <br />
        <Container gap="md" justify="center">
        <Input
          name="name"
          value={formValue.name}
          placeholder="Enter Name on Card"
          onChange={handleChange}
          boxStatus={boxStatus.name}
          testId="input-name"
        />
        <Input
          name="card"
          type="tel"
          value={formValue.card}
          placeholder="Enter Card Number"
          onChange={event => handleChange(event, 'card')}
          boxStatus={boxStatus.card}
          testId="input-card"
        />
        <div>
          <div>Card Expiration</div>
          <DropDown
            name="expMonth"
            options={monthsList}
            onChange={handleChange}
            boxStatus={boxStatus.exp}
            testId="input-month"
          />
          <DropDown
            name="expYear"
            options={yearsList}
            onChange={handleChange}
            boxStatus={boxStatus.exp}
            testId="input-year"
          />
        </div>
        <Input
          name="cvv"
          value={formValue.cvv}
          placeholder="Enter CVV"
          onChange={handleChange}
          boxStatus={boxStatus.cvv}
          testId="input-cvv"
        />
        <Button label="Submit Payment" variant="primary" onClick={event => handleSubmit(event)} />
        </Container>
      </Container>
    </form>
  )
}

export default PaymentForm;
