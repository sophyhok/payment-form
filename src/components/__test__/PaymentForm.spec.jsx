import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import PaymentForm from '../PaymentForm';

it('should match snapshot', () => {
  const { asFragment } = render(<PaymentForm />)
  expect(asFragment()).toMatchSnapshot();
});

describe('Validate Input: ', () => {

  it('should show error message when input name is invalid', () => {
    const { getByTestId, getByText } = render(<PaymentForm />);
    const button = getByTestId('button')
    const nameInput = getByTestId('input-name');

    fireEvent.change(nameInput, { target: { value: '123' } });
    fireEvent.click(button);

    expect(getByText('Name is invalid')).toBeInTheDocument();
  });
  it('should show error message when input card is invalid', () => {
    const { getByTestId, getByText } = render(<PaymentForm />);
    const button = getByTestId('button')
    const cardInput = getByTestId('input-card');

    fireEvent.change(cardInput, { target: { value: '0000' } });
    fireEvent.click(button);

    expect(getByText('Card is invalid')).toBeInTheDocument();
  });

  it('should show error message when input cvv is invalid', () => {
    const { getByTestId, getByText } = render(<PaymentForm />);
    const button = getByTestId('button')
    const cvvInput = getByTestId('input-cvv');

    fireEvent.change(cvvInput, { target: { value: '00' } });
    fireEvent.click(button);

    expect(getByText('CVV is invalid')).toBeInTheDocument();
  });

  it('should show error message when input date is invalid', () => {
    const { getByTestId, getByText } = render(<PaymentForm />);
    const button = getByTestId('button')
    const monthInput = getByTestId('input-month');
    const yearInput = getByTestId('input-year');

    fireEvent.change(monthInput, { target: { value: '01' } });
    fireEvent.change(yearInput, { target: { value: '2021' } });
    fireEvent.click(button);

    expect(getByText('Expiration is invalid')).toBeInTheDocument();
  });
});

