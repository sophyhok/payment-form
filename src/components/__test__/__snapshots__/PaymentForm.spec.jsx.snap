// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <form
    name="payment-form"
  >
    <div
      class="container container__align_center container__justify_center container__gap_none container__fontColor_dark"
      data-testid="container"
    >
      <h3>
        Enter your Credit Card information
      </h3>
      <div>
        <img
          alt="visa"
          src="visa.png"
        />
        <img
          alt="amex"
          src="amex.png"
        />
      </div>
      <br />
      <div
        class="container container__align_start container__justify_center container__gap_md container__fontColor_dark"
        data-testid="container"
      >
        <input
          class="input input__boxStatus_none"
          data-testid="input-name"
          name="name"
          pattern="[A-Za-z]{3}"
          placeholder="Enter Name on Card"
          type="text"
          value=""
        />
        <input
          class="input input__boxStatus_none"
          data-testid="input-card"
          name="card"
          pattern="[A-Za-z]{3}"
          placeholder="Enter Card Number"
          type="tel"
          value=""
        />
        <div>
          <div>
            Card Expiration
          </div>
          <select
            class="dropdown dropdown__boxStatus_none"
            data-testid="input-month"
            name="expMonth"
          >
            <option>
              01
            </option>
            <option>
              02
            </option>
            <option>
              03
            </option>
            <option>
              04
            </option>
            <option>
              05
            </option>
            <option>
              06
            </option>
            <option>
              07
            </option>
            <option>
              08
            </option>
            <option>
              09
            </option>
            <option>
              10
            </option>
            <option>
              11
            </option>
            <option>
              12
            </option>
          </select>
          <select
            class="dropdown dropdown__boxStatus_none"
            data-testid="input-year"
            name="expYear"
          >
            <option>
              2021
            </option>
            <option>
              2022
            </option>
            <option>
              2023
            </option>
            <option>
              2024
            </option>
            <option>
              2025
            </option>
            <option>
              2026
            </option>
            <option>
              2027
            </option>
            <option>
              2028
            </option>
            <option>
              2029
            </option>
            <option>
              2030
            </option>
            <option>
              2031
            </option>
          </select>
        </div>
        <input
          class="input input__boxStatus_none"
          data-testid="input-cvv"
          name="cvv"
          pattern="[A-Za-z]{3}"
          placeholder="Enter CVV"
          type="text"
          value=""
        />
        <button
          class="btn btn__primary"
          data-testid="button"
        >
          Submit Payment
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
